File: Code/Development/NeuralNet/src/matlab/Dataset.m
classdef Dataset < handle
    properties
        num_entries
        num_features
        dataset
    end
    methods
        function obj = Dataset(data, targets)
            obj.num_entries = size(data, 1);
            obj.num_features = size(data, 2);
            obj.dataset = [data, targets];
        end
        function data = get_data(obj)
            f = obj.num_features;
            data = obj.dataset(:,1:f);
        end
        function targets = get_targets(obj)
            f = obj.num_features;
            targets = obj.dataset(:,f+1:end);
        end
        function shuffle(obj)
            obj.dataset = obj.dataset(randperm(obj.num_entries),:);
        end
        function [dataset1, dataset2] = split(obj, ratio)
            n = floor(obj.num_entries * ratio);
            f = obj.num_features;
            dataset1 = Dataset(obj.dataset(1:n,1:f), obj.dataset(1:n,f+1:end));
            dataset2 = Dataset(obj.dataset(n+1:end,1:f), obj.dataset(n+1:end,f+1:end));
        end
        function classes = create_classes(obj, num_classes)
            classes = zeros(1, num_classes);
            targets = sort(obj.get_targets());
            batch_size = floor(obj.num_entries / num_classes);
            n_targets = batch_size * num_classes;
            j = 1;
            for i = 1:batch_size:n_targets
                batch = targets(i:i+batch_size);
                classes(j) = batch(1);
                j = j + 1;
            end
        end
        function t = encode_target(obj, target, classes)
            target_class = 0;
            n_classes = length(classes);
            for i = 1:n_classes
                if target >= classes(i)
                    target_class = i;
                end
            end
            t = zeros(1, n_classes);
            t(target_class) = 1.0;
        end
        function encode_targets(obj, classes)
            targets = obj.get_targets();
            enc_targets = zeros(obj.num_entries, length(classes));
            n_entries = obj.num_entries;
            for i = 1:n_entries
                enc_targets(i,:) = obj.encode_target(targets(i), classes);
            end
            obj.dataset = [obj.get_data(), enc_targets];
        end
        function batches = make_batches(obj, batch_size)
            n_entries = obj.num_entries - mod(obj.num_entries, batch_size);
            f = obj.num_features;
            batches = cell(1, n_entries / batch_size);
            j = 1;
            for i = 1:batch_size:n_entries
                batch_data = obj.dataset(i:i+batch_size,1:f);
                batch_targets = obj.dataset(i:i+batch_size,f+1:end);
                batches{j} = Dataset(batch_data, batch_targets);
                j = j + 1;
            end
        end
    end
end

File: Code/Development/NeuralNet/src/python/NeuralNetwork/Dataset.py
# -*- coding: utf-8 -*-
"""
Created on Fri Mar 10 09:55:15 2017

@author: Michael Austin Langford
"""

import numpy as np
import sys

class Dataset(object):
    def __init__(self, data, targets):
        self.__num_entries = data.shape[0]
        self.__num_features = data.shape[1]
        self.__dataset = np.hstack((data, targets))

    def get_data(self):
        return self.__dataset[:,:self.num_features]
    def set_data(self, v):
        self.__dataset[:,:self.num_features] = v
    data = property(fget=lambda self: self.get_data(),
                    fset=lambda self, v: self.set_data(v))
    
    def get_targets(self):
        return self.__dataset[:,self.num_features:]
    def set_targets(self, v):
        self.__dataset[:,self.num_features:] = v
    targets = property(fget=lambda self: self.get_targets(),
                       fset=lambda self, v: self.set_targets(v))
        
    def get_num_entries(self):
        return self.__num_entries
    num_entries = property(fget=lambda self: self.get_num_entries())
    
    def get_num_features(self):
        return self.__num_features
    num_features = property(fget=lambda self: self.get_num_features())

    def get_num_targets(self):
        return self.__dataset.shape[1] - self.__num_features
    num_targets = property(fget=lambda self: self.get_num_targets())

    def shuffle(self):
        np.random.shuffle(self.__dataset)

    def split(self, ratio):
        n = int(self.num_entries * ratio)
        dataset1 = Dataset(self.data[:n,:], self.targets[:n,:])
        dataset2 = Dataset(self.data[n:,:], self.targets[n:,:])
        return dataset1, dataset2

    def create_classes(self, num_classes):
        classes = np.zeros(num_classes)
        targets = np.sort(np.copy(self.targets), axis=0)
        batch_size = np.int(len(targets) / num_classes)
        num_targets = batch_size * num_classes
        j = 0
        for i in xrange(0, num_targets, batch_size):
            batch = targets[i:i+batch_size]
            classes[j] = batch[0]
            j += 1
        return classes

    def encode_targets(self, classes):
        targets = self.targets
        enc_targets = np.zeros((self.num_entries, classes.shape[0]))
        for i in xrange(self.num_entries):
            enc_targets[i,:] = self.encode_target(targets[i][0], classes)
        self.__dataset = np.hstack((self.data, enc_targets))
    
    def encode_target(self, target, classes):
        target_class = 0
        for j in xrange(len(classes)):
            if target >= classes[j]:
                target_class = j
        t = np.zeros((1, classes.shape[0]))
        t[0, target_class] = 1.0
        return t
    
    def make_batches(self, batch_size):
        batches = []
        for i in xrange(0, self.num_entries, batch_size):
            batch_data = self.data[i:i+batch_size,:]
            batch_targets = self.targets[i:i+batch_size,:]
            batches.append(Dataset(batch_data, batch_targets))
        return batches